java/drivers/driver-hazelcast4plus/src/main/java/com/hazelcast/simulator/tests/ucd/UCDTest.java

```diff
@@ -0,0 +1,87 @@
+package com.hazelcast.simulator.tests.ucd;
+
+import com.hazelcast.simulator.hz.HazelcastTest;
+import com.hazelcast.simulator.test.annotations.Setup;
+
+import java.net.URL;
+import java.net.URLClassLoader;
+
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Extendable test framework for Namespace performance testing.
+ * <p>
+ * Responsible for loading the UDF class with a URLClassLoader and
+ * optionally configuring a namespace with the loaded class.
+ * <p>
+ * Provides protected access to the loaded {@link #udf} for subclasses.
+ */
+public class UCDTest extends HazelcastTest {
+
+    /*
+   The directory where the UDF class can be found.
+   */
+    public String classDir;
+    /*
+    Fully qualified name of the class.
+     */
+    public String className;
+    /*
+    If namespaces should be enabled and the class loaded.
+    Note : it is required that the hazelcast.xml has a namespace
+    pre-configured.
+     */
+    public boolean useNamespace;
+    /*
+    The id of the namespace to add the class to. Must exist in the
+    hazelcast.xml.
+     */
+    public String namespaceId;
+    /*
+    The UDF class which is loaded with a URLClassLoader using the given
+    classDir and className test properties.
+     */
+    protected Class<?> udf;
+
+    /**
+     * Preloads UDF and configures namespace if required.
+     *
+     * @throws ClassNotFoundException if udf not found.
+     */
+    @Setup
+    public void setUp() throws ReflectiveOperationException {
+        loadUDF();
+        if (useNamespace) {
+            configureNamespace();
+        }
+    }
+
+    /**
+     * Loads the UDF class with a URL classloader.
+     *
+     * @throws ClassNotFoundException if udf not found.
+     */
+    @SuppressWarnings("resource")
+    private void loadUDF() throws ClassNotFoundException {
+        assertNotNull("classDir must be set", classDir);
+        assertNotNull("className must be set", className);
+        URL url = getClass().getClassLoader().getResource(classDir);
+        this.udf = new URLClassLoader(new URL[]{url}).loadClass(className);
+    }
+
+    /**
+     * Configures the loaded udf as a namespace with the configured
+     * namespacedId.
+     * <p>
+     * Note: this ID must be configured in the hazelcast.xml and should
+     * be applied to the hazelcast data structure in the hazelcast.xml.
+     */
+    private void configureNamespace() {
+        assertNotNull("namespaceId must be set", namespaceId);
+        assertNotNull("udf must be loaded", udf);
+//        NamespaceConfig nsc = new NamespaceConfig().setName(namespaceId).addClass(udf);
+//        targetInstance.getConfig().getNamespacesConfig()
+//                .addNamespaceConfig(nsc);
+        throw new IllegalArgumentException("UCD not supported in this version of Hazelcast!");
```

Comment:
Is this leftover debug?

Reply:
> Is this leftover debug?

No - depending on what version of Hazelcast you want to benchmark against, adjusted the programatic config available - i.e. on `5.3` you there is no `NamespaceConfig` so it wouldn't even compile, but when on `5.4` you want to make sure your using it.

It's possible to do this with reflection, but the code was very messy and I couldn't quite get it working.

Reply:
So the intention is that you uncomment the 3 lines currently commented, and comment out the exception throw, when you want to test UCN?

If that's the case we should document that clearly (in readme and code), and imo it should default to using the latest codebase approach (5.4, so the currently commented lines should be uncommented by default) - ideally we handle this properly, but I'm not sure what the usual process is for tests like this where we're using newly introduced code; perhaps @pveentjer has a good solution?

Reply:
> So the intention is that you uncomment the 3 lines currently commented, and comment out the exception throw, when you want to test UCN?

Yes

> If that's the case we should document that clearly (in readme and code)

It is [covered in the documentation](https://github.com/hazelcast/internal-benchmarks/blob/61686da88133a123c8cb7c0ee820cba24ff4c2ad/simulator2-benchmarks/user-code-namespaces/README.md#notes-for-running-the-tests), but I can see how it'd be easy to miss - I've update the code with comments as well - 698902a1a9c62a73da90644e53f2f6f82b30374d

> and imo it should default to using the latest codebase approach (5.4, so the currently commented lines should be uncommented by default) - ideally we handle this properly, but I'm not sure what the usual process is for tests like this where we're using newly introduced code; perhaps @pveentjer has a good solution?

It was originally done this way because UCN wasn't in `master`, but that's gone away now.

My concern is that if you ever wanted to compare some older version of Hazelcast - say Map throughput - you'd get a compile error for something unrelated. But I guess you could use that rationale to avoid adding *any* new Hazelcast functionality to the simulator, and that would be silly.